### ISSUE TITLE
fetch negotiation broken, fetches way more than it needs

### ISSUE THREAD

**Issue by arroz** [libgit2/libgit2#6532]
**TLDR:** The implementation of the negotiation protocol between client and server when fetching is broken in multiple ways. When a libgit2 client fetches from a server, this may cause the server to generate an extremely large pack containing mostly objects already present in the client.

**Detailed, long version:** During the fetch negotiation process, after receiving the full list of advertised references and respective OIDs from the server, the client sends a list of "want" commands with the requested OIDs, and "have" commands with a list of the commits it already has, in a way that minimized the amount of objects that need to be transfered.

According to core git's documentation, in a simplified way, the "haves" are generated by queuing the peeled commit OIDs of all the heads the client knows about, sorted by commit date (most recent at the head of the queue). Then, a loop is performed by popping the OID at the head of the queue, writing a "have" line for it, grabbing its parents, and inserting them on the queue, sorting them by commit date as well during the insertion. This loop is repeated at most 256 times, in batches of 20.

libgit2 is using the wrong order due to multiple bugs. Because of this, the OIDs used for the "have" commands are essentially arbitrary, and can be from commits years ago, instead of the most recent ones. When the server builds the packfile, it will assume those are the most recent the client has, and will send a huge amount of objects based on that wrong assumption.

### Reproduction steps

Reproducing this reliably is a bit hard due to the arbitrary nature of the problem (diagnostics show the order being used is the alphabetical order of the reference names, not the commit dates). But I've managed to find a situation that is reliably reproduced.

1. Setup a bare clone of the libgit2 repository in a location you can manipulate the file system directly. I used a VM with an HTTP server so I could easily inspect the traffic, but YMMV.

2. Replace the `packed-refs` file with the following one. This will bring the repository to how it was around March 16, 2023. [packed-refs-old.txt](https://github.com/libgit2/libgit2/files/11007351/packed-refs-old.txt)

3. Create two regular clones of the clone created in step 1.

4. In the clone created in step 1, replace `packed-refs` with the following version. This will add a few commits to two branches that were being worked on at the time, `ethomson/nonblocking` and `ethomson/schannel-2`. 
[packed-refs-new.txt](https://github.com/libgit2/libgit2/files/11007352/packed-refs-new.txt)

5. In one of the clones created on step 3, run core git command: `git fetch --all`. Note how much data it receives over the wire.

6. In the other clone created on step 3, use libgit2 to run a fetch operation, with a callback that prints the amount of received data. Note that it's almost 10 MB.

**Comment by arroz** [comment]
I have a fix for this, will submit in the next few days, since I'm trying to fix an additional problem.

**Comment by arroz** [comment]
@ethomson @carlosmn I split this into two PRs. The first one is simple and can go immediately if you're OK with it. The second one is a bit more problematic: although it works, it may increase the probability of timeout-related errors since we don't recover an HTTP keep-alive connection if it's closed. Can you read the comments on the second PR (#6540) and let me know your thoughts and a bit of guidance? Thank you! 😊

**Comment by ethomson** [comment]
Sorry I haven't had much time to think about this lately. I feel like we're still a bit stuck on what to do around timeouts. I'd like to think a bit more about the various work we have in flight but I did go ahead and merge the first PR. Thanks for this.
